Updating parallel by adding a branch causes a panic.

**Describe the bug**
From Slack:

https://knative.slack.com/archives/C9JP909F0/p1598546036094500

```
{"level":"info","ts":"2020-08-27T12:43:59.249Z","logger":"controller.knative.dev-eventing-pkg-reconciler-parallel.Reconciler","caller":"parallel/parallel.go:104","msg":"Reconciled Channel Object: default/broker-mgr-kn-parallel &{TypeMeta:{Kind:InMemoryChannel APIVersion:messaging.knative.dev/v1beta1} ObjectMeta:{Name:broker-mgr-kn-parallel GenerateName: Namespace:default SelfLink:/apis/messaging.knative.dev/v1beta1/namespaces/default/inmemorychannels/broker-mgr-kn-parallel UID:f46b2f66-6a1e-4e51-9ced-d3e2e7baf8ce ResourceVersion:67178 Generation:2 CreationTimestamp:2020-08-27 12:38:55 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[messaging.knative.dev/creator:system:serviceaccount:knative-eventing:eventing-controller messaging.knative.dev/lastModifier:system:serviceaccount:knative-eventing:eventing-controller messaging.knative.dev/subscribable:v1beta1] OwnerReferences:[{APIVersion:flows.knative.dev/v1 Kind:Parallel Name:broker-mgr UID:62914e91-384d-474e-8128-1454d3eac778 Controller:0xc001063b38 BlockOwnerDeletion:0xc001063b37}] Finalizers:[] ClusterName: ManagedFields:[{Manager:channel_dispatcher Operation:Update APIVersion:messaging.knative.dev/v1 Time:2020-08-27 12:38:57 +0000 UTC FieldsType:FieldsV1 FieldsV1:&FieldsV1{Raw:*[123 34 102 58 115 116 97 116 117 115 34 58 123 34 102 58 111 98 115 101 114 118 101 100 71 101 110 101 114 97 116 105 111 110 34 58 123 125 125 125],}} {Manager:controller Operation:Update APIVersion:messaging.knative.dev/v1beta1 Time:2020-08-27 12:38:57 +0000 UTC FieldsType:FieldsV1 FieldsV1:&FieldsV1{Raw:*[123 34 102 58 109 101 116 97 100 97 116 97 34 58 123 34 102 58 111 119 110 101 114 82 101 102 101 114 101 110 99 101 115 34 58 123 34 46 34 58 123 125 44 34 107 58 123 92 34 117 105 100 92 34 58 92 34 54 50 57 49 52 101 57 49 45 51 56 52 100 45 52 55 52 101 45 56 49 50 56 45 49 52 53 52 100 51 101 97 99 55 55 56 92 34 125 34 58 123 34 46 34 58 123 125 44 34 102 58 97 112 105 86 101 114 115 105 111 110 34 58 123 125 44 34 102 58 98 108 111 99 107 79 119 110 101 114 68 101 108 101 116 105 111 110 34 58 123 125 44 34 102 58 99 111 110 116 114 111 108 108 101 114 34 58 123 125 44 34 102 58 107 105 110 100 34 58 123 125 44 34 102 58 110 97 109 101 34 58 123 125 44 34 102 58 117 105 100 34 58 123 125 125 125 125 44 34 102 58 115 112 101 99 34 58 123 34 102 58 115 117 98 115 99 114 105 98 101 114 115 34 58 123 125 125 125],}} {Manager:channel_controller Operation:Update APIVersion:messaging.knative.dev/v1 Time:2020-08-27 12:38:58 +0000 UTC FieldsType:FieldsV1 FieldsV1:&FieldsV1{Raw:*[123 34 102 58 115 116 97 116 117 115 34 58 123 34 102 58 97 100 100 114 101 115 115 34 58 123 34 46 34 58 123 125 44 34 102 58 117 114 108 34 58 123 125 125 44 34 102 58 99 111 110 100 105 116 105 111 110 115 34 58 123 125 44 34 102 58 115 117 98 115 99 114 105 98 101 114 115 34 58 123 125 125 125],}}]} Spec:{SubscribableSpec:{Subscribers:[{UID:4eb15890-7f18-4e19-84fd-ee07c75e586b Generation:1 SubscriberURI:http://filter.customera.svc.cluster.local ReplyURI:http://broker-mgr-kn-parallel-0-kn-channel.default.svc.cluster.local Delivery:<nil>}]} Delivery:<nil>} Status:{Status:{ObservedGeneration:2 Conditions:[{Type:Addressable Status:True Severity: LastTransitionTime:{Inner:2020-08-27 12:38:55 +0000 UTC} Reason: Message:} {Type:ChannelServiceReady Status:True Severity: LastTransitionTime:{Inner:2020-08-27 12:38:55 +0000 UTC} Reason: Message:} {Type:DispatcherReady Status:True Severity: LastTransitionTime:{Inner:2020-08-27 12:38:55 +0000 UTC} Reason: Message:} {Type:EndpointsReady Status:True Severity: LastTransitionTime:{Inner:2020-08-27 12:38:55 +0000 UTC} Reason: Message:} {Type:Ready Status:True Severity: LastTransitionTime:{Inner:2020-08-27 12:38:55 +0000 UTC} Reason: Message:} {Type:ServiceReady Status:True Severity: LastTransitionTime:{Inner:2020-08-27 12:38:55 +0000 UTC} Reason: Message:}] Annotations:map[]} AddressStatus:{Address:0xc002849820} SubscribableStatus:{Subscribers:[{UID:4eb15890-7f18-4e19-84fd-ee07c75e586b ObservedGeneration:1 Ready:True Message:}]} DeadLetterChannel:<nil>}}","commit":"17179f5","knative.dev/traceid":"55e86eee-b37a-484f-b815-362f0ba60986","knative.dev/key":"default/broker-mgr"}
panic: runtime error: index out of range [1] with length 1
goroutine 394 [running]:
knative.dev/eventing/pkg/apis/flows/v1.(*ParallelStatus).PropagateChannelStatuses(0xc0014fe640, 0xc0011651e0, 0xc002114590, 0x2, 0x2)
	knative.dev/eventing/pkg/apis/flows/v1/parallel_lifecycle.go:161 +0x7de
knative.dev/eventing/pkg/reconciler/parallel.(*Reconciler).ReconcileKind(0xc000570640, 0x2181f60, 0xc0005f0180, 0xc0014fe500, 0x21477a0, 0xc0014fe640)
	knative.dev/eventing/pkg/reconciler/parallel/parallel.go:112 +0xa66
knative.dev/eventing/pkg/client/injection/reconciler/flows/v1/parallel.(*reconcilerImpl).Reconcile(0xc00057a280, 0x2181f60, 0xc0005f0150, 0xc00205bde0, 0x12, 0xc000b858c8, 0x2181f60)
	knative.dev/eventing/pkg/client/injection/reconciler/flows/v1/parallel/reconciler.go:242 +0x13cd
knative.dev/pkg/controller.(*Impl).processNextWorkItem(0xc000579e80, 0x0)
	knative.dev/pkg@v0.0.0-20200812224206-44c860147a87/controller/controller.go:511 +0x777
knative.dev/pkg/controller.(*Impl).RunContext.func3(0xc00203a0d0, 0xc000579e80)
	knative.dev/pkg@v0.0.0-20200812224206-44c860147a87/controller/controller.go:449 +0x53
created by knative.dev/pkg/controller.(*Impl).RunContext
	knative.dev/pkg@v0.0.0-20200812224206-44c860147a87/controller/controller.go:447 +0x1f5
```


A clear and concise description of what the bug is.

 **Expected behavior**
No panic.

 **To Reproduce**
Create parallel, add a branch to it, eventing-controller goes into crash loop.

 **Knative release version**

 **Additional context**
Add any other context about the problem here such as proposed priority

--- a/pkg/apis/flows/v1/parallel_lifecycle.go
+++ b/pkg/apis/flows/v1/parallel_lifecycle.go
@@ -77,7 +77,7 @@ func (ps *ParallelStatus) InitializeConditions() {
 // PropagateSubscriptionStatuses sets the ParallelConditionSubscriptionsReady based on
 // the status of the incoming subscriptions.
 func (ps *ParallelStatus) PropagateSubscriptionStatuses(filterSubscriptions []*messagingv1.Subscription, subscriptions []*messagingv1.Subscription) {
-	if ps.BranchStatuses == nil {
+	if ps.BranchStatuses == nil || len(subscriptions) != len(ps.BranchStatuses) {
 		ps.BranchStatuses = make([]ParallelBranchStatus, len(subscriptions))
 	}
 	allReady := true
@@ -136,7 +136,7 @@ func (ps *ParallelStatus) PropagateSubscriptionStatuses(filterSubscriptions []*m
 // PropagateChannelStatuses sets the ChannelStatuses and ParallelConditionChannelsReady based on the
 // status of the incoming channels.
 func (ps *ParallelStatus) PropagateChannelStatuses(ingressChannel *duckv1.Channelable, channels []*duckv1.Channelable) {
-	if ps.BranchStatuses == nil {
+	if ps.BranchStatuses == nil || len(channels) != len(ps.BranchStatuses) {
 		ps.BranchStatuses = make([]ParallelBranchStatus, len(channels))
 	}
 	allReady := true

--- a/pkg/apis/flows/v1/parallel_lifecycle_test.go
+++ b/pkg/apis/flows/v1/parallel_lifecycle_test.go
@@ -293,6 +293,37 @@ func TestParallelPropagateChannelStatuses(t *testing.T) {
 	}
 }
 
+func TestParallelPropagateChannelStatusUpdated(t *testing.T) {
+	inChannel := getChannelable(true)
+	initialChannels := []*eventingduckv1.Channelable{getChannelable(true)}
+	afterChannels := []*eventingduckv1.Channelable{getChannelable(true), getChannelable(true)}
+	ps := ParallelStatus{}
+	ps.PropagateChannelStatuses(inChannel, initialChannels)
+	if len(ps.BranchStatuses) != 1 {
+		t.Errorf("unexpected branchstatuses want 1 got %d", len(ps.BranchStatuses))
+	}
+	ps.PropagateChannelStatuses(inChannel, afterChannels)
+	if len(ps.BranchStatuses) != 2 {
+		t.Errorf("unexpected branchstatuses want 2 got %d", len(ps.BranchStatuses))
+	}
+}
+
+func TestParallelPropagateSubscriptionStatusUpdated(t *testing.T) {
+	initialFsubs := []*messagingv1.Subscription{getSubscription("fsub0", true)}
+	initialSubs := []*messagingv1.Subscription{getSubscription("sub0", true)}
+	afterFsubs := []*messagingv1.Subscription{getSubscription("fsub0", true), getSubscription("fsub1", true)}
+	afterSubs := []*messagingv1.Subscription{getSubscription("sub0", true), getSubscription("sub1", true)}
+	ps := ParallelStatus{}
+	ps.PropagateSubscriptionStatuses(initialFsubs, initialSubs)
+	if len(ps.BranchStatuses) != 1 {
+		t.Errorf("unexpected branchstatuses want 1 got %d", len(ps.BranchStatuses))
+	}
+	ps.PropagateSubscriptionStatuses(afterFsubs, afterSubs)
+	if len(ps.BranchStatuses) != 2 {
+		t.Errorf("unexpected branchstatuses want 2 got %d", len(ps.BranchStatuses))
+	}
+}
+
 func TestParallelReady(t *testing.T) {
 	tests := []struct {
 		name     string

